app/models/report.rb

```diff
@@ -3,15 +3,54 @@
 class Report < ApplicationRecord
   belongs_to :user
   has_many :comments, as: :commentable, dependent: :destroy
+  has_many :active_mentions, class_name: 'Mention', foreign_key: 'mentioning_id', inverse_of: 'mentioning', dependent: :destroy
+  has_many :passive_mentions, class_name: 'Mention', foreign_key: 'mentioned_id', inverse_of: 'mentioned', dependent: :destroy
+  has_many :mentioning_reports, through: :active_mentions, source: :mentioned
+  has_many :mentioned_reports, through: :passive_mentions, source: :mentioning
 
   validates :title, presence: true
   validates :content, presence: true
 
+  after_save :save_mentions
+
   def editable?(target_user)
     user == target_user
   end
 
   def created_on
     created_at.to_date
   end
+
+  private
+
+  def save_mentions
+    mentioning_report_ids_diff = fetch_mentioning_report_ids_diff
+    add_ids = mentioning_report_ids_diff[:added_ids]
+    remove_ids = mentioning_report_ids_diff[:removed_ids]
```

Comment:
`fetch_mentioning_report_ids_diff` で配列を返すようにして、

```suggestion
    add_ids, remove_ids = fetch_mentioning_report_ids_diff
```

としてしまうのもありかもしれません。

（戻り値が多いとハッシュの方がいいですが、2〜3個までなら配列でもいいかな、という感じ）